parachain/src/wasm_executor/mod.rs

```diff
@@ -84,372 +82,169 @@ pub enum Error {
 	External(String),
+	#[display(fmt = "Shared memory error: {}", _0)]
+	#[cfg(not(target_os = "unknown"))]
+	SharedMem(shared_memory::SharedMemError),
 }
 
 impl std::error::Error for Error {
 	fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
 		match self {
-			Error::Wasm(ref err) => Some(err),
-			Error::Externalities(ref err) => Some(err),
+			Error::WasmExecutor(ref err) => Some(err),
 			Error::Io(ref err) => Some(err),
 			Error::System(ref err) => Some(&**err),
+			#[cfg(not(target_os = "unknown"))]
+			Error::SharedMem(ref err) => Some(err),
 			_ => None,
 		}
 	}
 }
 
-/// Errors that can occur in externalities of parachain validation.
-#[derive(Debug, Clone)]
-pub enum ExternalitiesError {
-	/// Unable to post a message due to the given reason.
-	CannotPostMessage(&'static str),
-}
-
 /// Externalities for parachain validation.
-pub trait Externalities {
+pub trait Externalities: Send {
 	/// Called when a message is to be posted to another parachain.
-	fn post_message(&mut self, message: MessageRef) -> Result<(), ExternalitiesError>;
+	fn post_message(&mut self, message: TargetedMessage) -> Result<(), String>;
 
 	/// Called when a message is to be posted to the parachain's relay chain.
-	fn post_upward_message(&mut self, message: UpwardMessageRef) -> Result<(), ExternalitiesError>;
+	fn post_upward_message(&mut self, message: UpwardMessage) -> Result<(), String>;
 }
 
-impl fmt::Display for ExternalitiesError {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		match *self {
-			ExternalitiesError::CannotPostMessage(ref s)
-				=> write!(f, "Cannot post message: {}", s),
-		}
+/// Validate a candidate under the given validation code.
+///
+/// This will fail if the validation code is not a proper parachain validation module.
+pub fn validate_candidate<E: Externalities + 'static>(
+	validation_code: &[u8],
+	params: ValidationParams,
+	ext: E,
+	options: ExecutionMode,
+) -> Result<ValidationResult, Error> {
+	match options {
+		ExecutionMode::Local => {
+			validate_candidate_internal(validation_code, &params.encode(), ext)
+		},
+		#[cfg(not(target_os = "unknown"))]
+		ExecutionMode::Remote => {
+			validation_host::validate_candidate(validation_code, params, ext, false)
+		},
+		#[cfg(not(target_os = "unknown"))]
+		ExecutionMode::RemoteTest => {
+			validation_host::validate_candidate(validation_code, params, ext, true)
+		},
+		#[cfg(target_os = "unknown")]
+		ExecutionMode::Remote =>
+			Err(Error::System("Remote validator not available".to_string().into())),
+		#[cfg(target_os = "unknown")]
+		ExecutionMode::RemoteTest =>
+			Err(Error::System("Remote validator not available".to_string().into())),
 	}
 }
 
-impl wasmi::HostError for ExternalitiesError {}
-impl std::error::Error for ExternalitiesError {}
+/// The host functions provided by the wasm executor to the parachain wasm blob.
+type HostFunctions = (sp_io::SubstrateHostFunctions, crate::wasm_api::parachain::HostFunctions);
```

Comment:
Is this exposing all of Substrate's host functions to the parachain WASM blob? incl. storage etc?

Reply:
I see a little lower down that these are all implemented but as panicking stubs. It's really weird to make the parachain WASM link to all of these and then have it never call them.

Reply:
Yes it is exposing all of them. 

Most of the stuff is related to the current design of Substrate and its externalities. 

I would really like to not expose the storage functions from the host, but that also requires changes to the wasm executor. The executor currently tries to resolve all extern functions and otherwise returns an error. In my opinion it should ignore functions that it does not know and error when the blob tries to execute them.

What kind of functions to we want to expose to the parachain wasm blob? I think I asked this before and the answer was that we expose the same functionality as Substrate does. So crypto functions etc. If we have support for ignoring extern functions, we can make this more fine grained, but I don't want to block on this for now.